conn.go

```diff
@@ -801,139 +803,210 @@ type callResp struct {
+	case c.semaphore <- struct{}{}:
+		// acquired
+	}
+
+	defer func() {
+		// release
+		<-c.semaphore
+	}()
+
 	if c.timeout > 0 {
-		c.w.SetWriteDeadline(time.Now().Add(c.timeout))
+		err := c.w.SetWriteDeadline(time.Now().Add(c.timeout))
+		if err != nil {
+			return 0, err
+		}
 	}
 	return c.w.Write(p)
 }
 
-func newWriteCoalescer(conn net.Conn, timeout time.Duration, d time.Duration, quit <-chan struct{}) *writeCoalescer {
+func newWriteCoalescer(conn deadlineWriter, writeTimeout, coalesceDuration time.Duration,
+	quit <-chan struct{}) *writeCoalescer {
 	wc := &writeCoalescer{
-		writeCh: make(chan struct{}), // TODO: could this be sync?
-		cond:    sync.NewCond(&sync.Mutex{}),
+		writeCh: make(chan writeRequest),
 		c:       conn,
 		quit:    quit,
-		timeout: timeout,
+		timeout: writeTimeout,
 	}
-	go wc.writeFlusher(d)
+	go wc.writeFlusher(coalesceDuration)
 	return wc
 }
 
 type writeCoalescer struct {
-	c net.Conn
+	c deadlineWriter
+
+	mu sync.Mutex
 
 	quit    <-chan struct{}
-	writeCh chan struct{}
-	running bool
+	writeCh chan writeRequest
 
-	// cond waits for the buffer to be flushed
-	cond    *sync.Cond
-	buffers net.Buffers
 	timeout time.Duration
 
-	// result of the write
-	err error
-}
-
-func (w *writeCoalescer) flushLocked() {
-	w.running = false
-	if len(w.buffers) == 0 {
-		return
-	}
-
-	if w.timeout > 0 {
-		w.c.SetWriteDeadline(time.Now().Add(w.timeout))
-	}
-
-	// Given we are going to do a fanout n is useless and according to
-	// the docs WriteTo should return 0 and err or bytes written and
-	// no error.
-	_, w.err = w.buffers.WriteTo(w.c)
-	if w.err != nil {
-		w.buffers = nil
-	}
-	w.cond.Broadcast()
+	testEnqueuedHook func()
+	testFlushedHook  func()
 }
 
-func (w *writeCoalescer) flush() {
-	w.cond.L.Lock()
-	w.flushLocked()
-	w.cond.L.Unlock()
+type writeRequest struct {
+	// resultChan is a channel (with buffer size 1) where to send results of the write.
+	resultChan chan<- writeResult
+	// data to write.
+	data []byte
 }
 
-func (w *writeCoalescer) stop() {
-	w.cond.L.Lock()
-	defer w.cond.L.Unlock()
-
-	w.flushLocked()
-	// nil the channel out sends block forever on it
-	// instead of closing which causes a send on closed channel
-	// panic.
-	w.writeCh = nil
+type writeResult struct {
+	n   int
+	err error
 }
 
-func (w *writeCoalescer) Write(p []byte) (int, error) {
-	w.cond.L.Lock()
-
-	if !w.running {
-		select {
-		case w.writeCh <- struct{}{}:
-			w.running = true
-		case <-w.quit:
-			w.cond.L.Unlock()
-			return 0, io.EOF // TODO: better error here?
-		}
+// writeContext implements contextWriter.
+func (w *writeCoalescer) writeContext(ctx context.Context, p []byte) (int, error) {
+	resultChan := make(chan writeResult, 1)
+	wr := writeRequest{
+		resultChan: resultChan,
+		data:       p,
 	}
 
-	w.buffers = append(w.buffers, p)
-	for len(w.buffers) != 0 {
-		w.cond.Wait()
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case <-w.quit:
+		return 0, io.EOF // TODO: better error here?
```

Comment:
**nitpick**: forgotten TODO? TBH I have no idea what would be better 🤷

Reply:
It was in the original code, so I kept it there. Another option is `ErrClosedConnection`.